syncstorage/tests/functional/test_storage.py

```diff
@@ -1509,24 +1515,87 @@ def test_batch_size_limits(self):
         self.assertTrue('max_total_records' in limits)
         self.assertTrue('max_total_bytes' in limits)
         self.assertTrue('max_record_payload_bytes' in limits)
+        self.assertTrue('max_request_bytes' in limits)
 
         endpoint = self.root + '/storage/col2?batch=true'
 
+        # There are certain obvious constraints on these limits,
+        # violations of which would be very confusing for clients.
+
+        self.assertTrue(
+            limits['max_request_bytes'] > limits['max_post_bytes']
+        )
+        self.assertTrue(
+            limits['max_post_bytes'] >= limits['max_record_payload_bytes']
+        )
+        self.assertTrue(
+            limits['max_total_records'] >= limits['max_post_records']
+        )
+        self.assertTrue(
+            limits['max_total_bytes'] >= limits['max_post_bytes']
+        )
+
+        # `max_post_records` is an (inclusive) limit on
+        # the number of items in a single post.
+
+        res = self.app.post_json(endpoint, [], headers={
+          'X-Weave-Records': str(limits['max_post_records'])
+        })
+        self.assertFalse(res.json['failed'])
         res = self.app.post_json(endpoint, [], headers={
           'X-Weave-Records': str(limits['max_post_records'] + 1)
         }, status=400)
         self.assertEquals(res.json, WEAVE_SIZE_LIMIT_EXCEEDED)
 
+        bsos = [{'id': str(x), 'payload': ''}
+                for x in xrange(limits['max_post_records'])]
+        res = self.app.post_json(endpoint, bsos)
+        self.assertFalse(res.json['failed'])
+        bsos.append({'id': 'toomany', 'payload': ''})
+        res = self.app.post_json(endpoint, bsos)
+        self.assertEquals(res.json['failed']['toomany'], 'retry bso')
+
+        # `max_total_records` is an (inclusive) limit on the
```

Comment:
I don't think desktop will say this. Maybe depend is the wrong word then? (Or we need a client fix...). Same for `X-Weave-Total-Bytes` below.

Reply:
"We can only enforce it if..." is probably a more accurate phrasing.  Of course iif desktop *can* send this it would be helpful :-)